src/main/java/picard/fingerprint/FingerprintChecker.java

```diff
@@ -679,6 +728,76 @@ public Map<FingerprintIdDetails, Fingerprint> fingerprintFiles(final Collection<
         return retval;
     }
 
+    /**
+     * Same as previous method, but allows a Path -> Path map as input, to manually point input files to corresponding index.
+     */
+    public Map<FingerprintIdDetails, Fingerprint> fingerprintFiles(final Collection<Path> files, final Map<Path, Path> indexPathMap,
+                                                                   final int threads, final int waitTime, final TimeUnit waitTimeUnit) {
+
+        // Generate fingerprints from each file
+        final AtomicInteger filesRead = new AtomicInteger(0);
+
+        final ExecutorService executor = new ThreadPoolExecutorWithExceptions(threads);
+        final ExecutorCompletionService<Path> executorCompletionService = new ExecutorCompletionService<>(executor);
+        final Map<FingerprintIdDetails, Fingerprint> retval = new ConcurrentHashMap<>(files.size());
+
+        for (final Path p : files) {
+            executorCompletionService.submit(() -> {
+
+                final Map<FingerprintIdDetails, Fingerprint> oneFileFingerprints;
+                log.debug("Processed file: " + p.toUri().toString() + " (" + filesRead.get() + ")");
+
+                if (CheckFingerprint.fileContainsReads(p)) {
+                    if (indexPathMap.containsKey(p)) {
```

Comment:
This might be cleaner to add to the method above rather than copying the surrounding code. You can check if `indexPathMap` is null to keep the "default index file" option.

Reply:
Refactored a bit to implement this